### PR TITLE
feat: add `npm_config_user_agent` support

### LINKED ISSUE #19: Add support for detecting package manager when no lock file is present
### Clear and concise description of the problem

In [remix-run/remix](https://github.com/remix-run/remix), we have our own implementation of `detectPackageManager`.
Only difference is that we currently check the `npm_config_user_agent` to determine which package manager is used.
https://github.com/remix-run/remix/blob/8649f1a561ce5ea0bfba5ab862c0c94052f4a5c3/packages/remix-dev/cli/detectPackageManager.ts#L10-L23

We do this because people can create a new project with one of our so called ["Stacks"](https://remix.run/docs/en/main/guides/templates#stacks), which can potentially _**not**_ have a lock file.

We also have a custom implementation in each of our stacks for the commands that we use:
- https://github.com/remix-run/indie-stack/blob/e63be8d06c20f92ffbcf30d8c69bf34258f26483/remix.init/index.js#L24-L64
- https://github.com/remix-run/blues-stack/blob/b050c907bc2a2778c9b4d9437d89c7c4751194fb/remix.init/index.js#L24-L64
- https://github.com/remix-run/grunge-stack/blob/fcd4729a209963ae1dd912e2d2cf0fdcd8644849/remix.init/index.js#L11-L47

It would be awesome if we could get rid of the custom implementation and use `package-manager-detector` instead.

### Suggested solution

Make it possible to detect package manager using `npm_config_user_agent`

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/antfu/.github/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://github.com/antfu/contribute).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by userquin**
### Linked Issues

resolves #19

### CODE REVIEW

**Comment by userquin** [comment]
Why breaking? The default flag is undefined and will not check for the env entry

File: src/types.ts [hunk 1]
```diff
@@ -42,6 +42,13 @@ export interface DetectOptions {
    * @param packageManager - The `packageManager` value from package.json file.
    */
   onUnknown?: (packageManager: string) => DetectResult | null | undefined
+  /**
+   * Whether to include `npm_config_user_agent` check when running in another process.
+   *
+   * When set to `true`, `detect/detectSync` will check only for `npm_config_user_agent`.
+   * When set to `'continue'`, `detect/detectSync` will check for `npm_config_user_agent`, if missing, will proceed with lock and package.json checks.
+   */
+  npm_config_user_agent?: true | 'continue'
```
Comment (userquin) [inline src/types.ts:51]: Use true only

**Comment by benmccann** [comment]
Thanks for the update! Yeah, not a breaking change

Though I do think separate methods would be better for tree shaking. Right now we basically have `if (opt) { method1() } else { method2() }`. Why not just let the user call the method they need directly to avoid including code for the method they don't need?

File: src/detect.ts [hunk 2]
```diff
@@ -61,6 +81,32 @@ export function detectSync(options: DetectOptions = {}): DetectResult | null {
   return null
 }
 
+function processUserAgent(pm: any, onUnknown: DetectOptions['onUnknown']): DetectResult | null {
```
Comment (benmccann) [inline src/detect.ts:84]: we should probably add some tests for this

here's an implementation from another project I've worked on:
```
export function getUserAgent(): AgentName | undefined {
	const userAgent = process.env.npm_config_user_agent;
	if (!userAgent) return undefined;

	const pmSpec = userAgent.split(' ')[0]!;
	const separatorPos = pmSpec.lastIndexOf('/');
	const name = pmSpec.substring(0, separatorPos) as AgentName;
	return AGENTS.includes(name) ? name : undefined;
}
```

i'm surprised the two implementations are so different, but haven't looked into the differences or if one is handling some cases differently

**Comment by MichaelDeBoey** [comment]
Now that I created https://github.com/antfu-collective/package-manager-detector/issues/21, I'm not sure if using `npm_config_user_agent` would (always) be the correct thing to look at though 🤔

**Review by benmccann — APPROVED** [review]
nice!!

Thanks @userquin !

**Comment by benmccann** [comment]
> Now that I created https://github.com/antfu-collective/package-manager-detector/issues/21, I'm not sure if using npm_config_user_agent would (always) be the correct thing to look at though 🤔

We can add support for Deno in this method when implementing https://github.com/antfu-collective/package-manager-detector/issues/21. I'll comment there with some more details. I don't think it should block this PR though

File: src/detect.ts [hunk 3]
```diff
@@ -61,6 +63,20 @@ export function detectSync(options: DetectOptions = {}): DetectResult | null {
...
 
+/**
+ * Detects the package manager used in the running process.
+ *
+ * This method will check for `process.env.npm_config_user_agent`.
+ */
+export function getUserAgent(): AgentName | null {
+  const userAgent = process.env.npm_config_user_agent
+  if (!userAgent)
+    return null
```
Comment (MichaelDeBoey) [inline src/detect.ts:74]: This makes it a bit more readable for less experienced people

```suggestion
  if (!userAgent) {
    return null
  }
```

File: src/detect.ts [hunk 4]
```diff
@@ -61,6 +63,20 @@ export function detectSync(options: DetectOptions = {}): DetectResult | null {
...
+
+  const name = userAgent.split(' ')[0].split('/')[0] as AgentName
```
Comment (MichaelDeBoey) [inline src/detect.ts:76]: I don't think we first need to split on empty space (`' '`) and we can just split on `/` directly, just like we do in the Remix codebase?
https://github.com/remix-run/remix/blob/8649f1a561ce5ea0bfba5ab862c0c94052f4a5c3/packages/remix-dev/cli/detectPackageManager.ts#L10-L23

```suggestion
  const name = userAgent.split('/')[0] as AgentName
```

File: test/user-agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import { getUserAgent } from '../src/detect'
+
+describe('get user agent', () => {
+  beforeEach(() => {
+    vi.unstubAllEnvs()
+  })
+  ;[
+    ['npm', 'npm/10.9.0 node/v20.17.0 linux x64 workspaces/false'],
+    ['yarn', 'yarn/1.22.11 npm/? node/v14.17.6 darwin x64'],
+    ['pnpm', 'pnpm/9.12.1 npm/? node/v20.17.0 linux x64'],
```
Comment (MichaelDeBoey) [inline test/user-agent.spec.ts:11]: Maybe we should add a test for bun as well?

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

File: test/user-agent.spec.ts (hunk 5 above)
Comment (benmccann) [inline test/user-agent.spec.ts:11]:
```suggestion
    ['pnpm', 'pnpm/9.12.1 npm/? node/v20.17.0 linux x64'],
    ['bun', 'bun/1.1.8 npm/? node/v21.6.0 linux x64'],
```

**Review by MichaelDeBoey — APPROVED** [review]

**Review by antfu — APPROVED** [review]

**Pull request merged**